com.unity.sg2/Editor/GraphUI/EditorCommon/Preview/PreviewUpdateDispatcher.cs

```diff
@@ -0,0 +1,168 @@
+        {
+            m_GraphModel = shaderGraphModel;
+            m_MainPreviewData = shaderGraphModel.MainPreviewData;
+
+            m_Scheduler = scheduler;
+
+            m_TimeDependentNodes = new();
+            m_OwningWindowReference = owningWindow;
+
+            // Initialize the headless preview
+            m_PreviewHandlerInstance = new HeadlessPreviewManager();
+            m_PreviewHandlerInstance.Initialize(shaderGraphModel.DefaultContextName, m_MainPreviewData.MainPreviewSize);
+            m_PreviewHandlerInstance.SetActiveGraph(shaderGraphModel.GraphHandler);
+            m_PreviewHandlerInstance.SetActiveRegistry(shaderGraphModel.RegistryInstance.Registry);
+            m_PreviewHandlerInstance.SetActiveTarget(shaderGraphModel.ActiveTarget);
+            m_PreviewHandlerInstance.SetPreviewUpdateReceiver(previewUpdateReceiver);
+
+            // Request preview data for any nodes that exist on graph load
+            // TODO: Think about how to auto-persist this so we don't need to recalculate every time
+            foreach (var nodeModel in m_GraphModel.NodeModels)
+            {
+                switch (nodeModel)
+                {
+                    case GraphDataContextNodeModel contextNode when contextNode.IsMainContextNode():
+                        m_PreviewHandlerInstance.RequestPreviewUpdate(contextNode.graphDataName, m_Scheduler, contextNode.NodePreviewMode);
+                        break;
+                    case GraphDataNodeModel graphDataNodeModel when graphDataNodeModel.HasPreview:
+                        m_PreviewHandlerInstance.RequestPreviewUpdate(graphDataNodeModel.graphDataName, m_Scheduler, graphDataNodeModel.NodePreviewMode);
+                        if(m_GraphModel.DoesNodeRequireTime(graphDataNodeModel))
+                            m_TimeDependentNodes.Add(graphDataNodeModel.graphDataName);
+                        break;
+
+
+                }
+            }
+        }
+
+        public void OnListenerAdded(string listenerID, HeadlessPreviewManager.PreviewRenderMode previewRenderMode, bool isListenerTimeDependent)
+        {
+            if (isListenerTimeDependent)
+                m_TimeDependentNodes.Add(listenerID);
+
+            m_PreviewHandlerInstance.RequestPreviewUpdate(listenerID, m_Scheduler, previewRenderMode);
+        }
+
+        public void OnListenerConnectionChanged(string listenerID, bool wasNodeDeleted = false)
+        {
+            var impactedNodes = m_PreviewHandlerInstance.NotifyNodeFlowChanged(listenerID, wasNodeDeleted);
+
+            if (m_GraphModel.DoesNodeRequireTime(listenerID))
+                m_TimeDependentNodes.Add(listenerID);
+
+            // If a node was deleted we don't want to issue update calls for it thereafter
+            if (wasNodeDeleted)
+            {
+                impactedNodes.Remove(listenerID);
+                m_TimeDependentNodes.Remove(listenerID);
+            }
+
+            foreach (var downstreamNode in impactedNodes)
```

Comment:
Can we move evaluation of halo into the preview service? so that update dispatcher isnt responsible